src/views/PositionPdfPlot/PositionPdfPlotWidget.tsx

```diff
@@ -0,0 +1,229 @@
+        const i1 = Math.max(0, Math.floor((visibleTimeStartSeconds - startTimeSec) * samplingFrequency))
+        const i2 = Math.min(dataModel.numTimepoints, Math.ceil((visibleTimeEndSeconds - startTimeSec) * samplingFrequency))
+        if (i2 <= i1) return {visibleValues: undefined, t1: 0, t2: 0}
+        let downsampleFactor = 1
+        while ((i2 - i1) / downsampleFactor > 1000) {
+            downsampleFactor *= dataModel.multiscaleFactor
+        }
+        const j1 = Math.floor(i1 / downsampleFactor)
+        const j2 = Math.ceil(i2 / downsampleFactor)
+        const visibleValues = dataModel.get(j1, j2, downsampleFactor)
+        const t1 = startTimeSec + j1 * downsampleFactor / samplingFrequency
+        const t2 = startTimeSec + j2 * downsampleFactor / samplingFrequency
+        return {visibleValues, t1, t2}
+    }, [dataModel, visibleTimeStartSeconds, visibleTimeEndSeconds, startTimeSec, samplingFrequency])
+
+    const panelCount = 1
+    const { panelWidth, panelHeight } = usePanelDimensions(width, height, panelCount, panelSpacing, margins)
+    const pixelsPerSecond = usePixelsPerSecond(panelWidth, visibleTimeStartSeconds, visibleTimeEndSeconds)
+    const timeToPixelMatrix = use1dTimeToPixelMatrix(pixelsPerSecond, visibleTimeStartSeconds)
+
+    const pixelTimes = useMemo(() => {
+        const augmentedVisibleTimesMatrix = matrix([ [t1, t2], new Array(2).fill(1) ])
+        return multiply(timeToPixelMatrix, augmentedVisibleTimesMatrix).valueOf() as number[]
+    }, [t1, t2, timeToPixelMatrix])
+
+    const {minValue, maxValue} = useMemo(() => {
+        if (!visibleValues) return {minValue: 0, maxValue: 0}
+        return {
+            minValue: min(visibleValues.map(a => (min(a)))),
+            maxValue: max(visibleValues.map(a => (max(a)))),
+        }
+    }, [visibleValues])
+
+    const imageData = useMemo(() => {
+        if (!visibleValues) return undefined
+        if (minValue === undefined) return undefined
+        if (maxValue === undefined) return undefined
+        const N1 = visibleValues.length
+        const N2 = visibleValues[0].length
+        const imageData = new ImageData(N1, N2)
+        let i = 0
+        for (let i2=0; i2<N2; i2++) {
+            for (let i1=0; i1<N1; i1++) {
+                const vv = visibleValues[i1][N2 - 1 - i2]
+                if (vv !== undefined) {
+                    const v = (vv - minValue) / (maxValue - minValue)
+                    const col = colorForValue(v)
+                    imageData.data[i + 0] = col[0]
+                    imageData.data[i + 1] = col[1]
+                    imageData.data[i + 2] = col[2]
+                    imageData.data[i + 3] = 255
+                }
+                else {
+                    imageData.data[i + 0] = 0
+                    imageData.data[i + 1] = 0
+                    imageData.data[i + 2] = 0
+                    imageData.data[i + 3] = 0
+                }
+                i += 4
+            }
+        }
```

Comment:
This is a bit terse. Looks like you're making a heatmap by converting the vector components of the underlying pdf data object (as downsampled) into a proportion of their range, and then writing that into the RGB channels, something like that?

Using `i2` for the outer loop iterator and `i1` for the inner loop iterator is not the most conventional choice :)
The outer loop is iterating over which component of the pdf you're encoding, and the inner one is iterating over the time slot, do I have that right? I'm assuming this has to do with the layout of the `ImageData` in memory and how it corresponds to the rows/columns of pixels in the underlying image object?

Reply:
Yes. The reason the loops are nested the way they are is because we need to write imageData in a specific order.